aten/src/TH/generic/THTensorMath.c

```diff
@@ -3629,22 +3629,32 @@ TENSOR_IMPLEMENT_LOGICAL(ne,!=)
     ptrdiff_t r_Size = THTensor_(nElement)(r_);               \
     int r_Contig = THTensor_(isContiguous)(r_);               \
     int tContig = THTensor_(isContiguous)(t);                 \
-    int inOMP = omp_in_parallel();                            \
-    if( (r_Size > TH_OMP_OVERHEAD_THRESHOLD) && (!inOMP) ){   \
-      TH_TENSOR_APPLY2_OMP(r_Size, r_Contig, tContig, real, r_, real, t, *r__data = CFUNC(*t_data););        \
-    }                                                                                                        \
-    else {                                                                                                   \
-      TH_TENSOR_APPLY2(real, r_, real, t, *r__data = CFUNC(*t_data););                                       \
-    }                                                                                                        \
+    if (r_Contig && tContig) {                                \
+      TH_TENSOR_APPLY2_CONTIG(real, r_, real, t, THVector_(NAME)(r__data, t_data, r__len););                   \
+    } else {                                                                                                   \
+      int inOMP = omp_in_parallel();                            \
+      if( (r_Size > TH_OMP_OVERHEAD_THRESHOLD) && (!inOMP) ){   \
+        TH_TENSOR_APPLY2_OMP(r_Size, r_Contig, tContig, real, r_, real, t, *r__data = CFUNC(*t_data););        \
```

Comment:
I am concerned about modifying this macro for all the non-AVX functions just so that sigmoid gets remapped. Is it possible to write this so _just_ the functions we want (for now, just sigmoid) check and dispatch to their vector equivalent?  @gchanan do you know why TH_TENSOR_APPLY2_OMP no longer uses TH_TENSOR_APPLY2_CONTIG by default?

Reply:
@zdevito `TH_TENSOR_APPLY2_OMP` was recently introduced and is a superset of `TH_TENSOR_APPLY2_CONTIG`, and handles the case where the inputs are non-contiguous as well.
I believe we can now remove `TH_TENSOR_APPLY2_CONTIG` as this case is covered by the `*_OMP` macros.
Also, as it fuses the contiguous dimensions, the overhead of using it compared to `*CONTIG` should be minimal I think.

But, if we remove the `*CONTIG` macros, I think we would need to pay attention to the fact that 
[`TH_TENSOR_APPLY2_OMP` tries to force vectorization at the compiler level for contiguous tensors](https://github.com/pytorch/pytorch/blob/master/aten/src/TH/THTensorApply.h#L372-L388), which might make the binaries that we release not work on cpus without vector instructions, as it would lack the runtime dispatch.

Reply:
> TH_TENSOR_APPLY2_OMP tries to force vectorization at the compiler level for contiguous tensors

This does not matter. We set AVX/AVX2 flags only for particular C files. So even if compiler tries to vectorize other files, it wont use AVX/AVX2 intrinsics.

Reply:
The code as written seems to _disable_ vectorization by bypassing THVector_ methods, which is why @vedanuj needs to route around it to get sigmoid to vectorize. That seems bad when we have a custom vectorized implementation, since we will fallback to a worse version. @fmassa Do you have a idea of how to fix these cases so that they can work together? Ideally we fallback to trying to force the compiler to vectorize only when we don't already have a custom vectorized implementation.  The compiler is not going to be able to vectorize sigmoid since it doesn't have a vectorized exp.

Reply:
@zdevito I can make these changes into a new macro for vectorized functions(for now only sigmoid will use that) and any other functions that will be vectorized later can use this new macro. However changing the LAB_IMPLEMENT_BASIC_FUCNTION macro doesn't hurt  perf for the other non vectorized functions because their THVector implementations are calling the serialized version of the functions in [` VECTOR_IMPLEMENT_FUNCTION`](https://github.com/pytorch/pytorch/blob/master/aten/src/TH/generic/THVectorDefault.cpp#L185). I tested a few functions like tanh, log, exp and didn't notice any degradation in their perf with this change.